### ISSUE TITLE
Add hive metastore catalog support

### ISSUE THREAD

**Issue by Xuanwo** [apache/iceberg-rust#113]
## Tasks

- [x] https://github.com/apache/iceberg-rust/pull/112
- [ ] Implement catalog trait for hms

**Comment by marvinlanhenke** [comment]
@Xuanwo I've started looking into this (just started yesterday taking a look at the codebase in general) - how would you go about unit testing the catalog trait for hms? The integration test could probably be done using the docker-compose utils etc. but I am not sure how to best unit test this? Do you have any suggestions or pointers for me to look at? Thank you very much and best regards.

**Comment by Xuanwo** [comment]
> I've done a simple prototype with my hive-metastore running locally via docker-compose, which works just fine. So testing via integration tests seems to be no problem.

It's fine to have integration tests. It doesn't make sense to me to mock HMS.

**Comment by marvinlanhenke** [comment]
@Xuanwo 
Yes, I thought the same - mocking the HMS would not be beneficial.
I setup the basis test-infra for the hms-catalog. In the next couple of days I will try to work on / finish the implementation of the catalog.

**Comment by Xuanwo** [comment]
> I setup the basis test-infra for the hms-catalog. In the next couple of days I will try to work on / finish the implementation of the catalog.

That's great!

**Comment by marvinlanhenke** [comment]
@Xuanwo @liurenjie1024
...just a quick update and a request for feedback.

I 'finished' the work on all the namespace operations. 
Since this would be my first contribution to this repo, I would appreciate it if someone could take a quick look at the current status. 

I just want to make sure it aligns with the rest of the codebase, before I put in the effort to implement the rest of the catalog.
https://github.com/marvinlanhenke/iceberg-rust/blob/hms/crates/catalog/hms/src/catalog.rs#L260-L428

sidenote:
For the hive-metastore docker-compose, I used a file I had from a previous project; however this needs some optimization since the connection to the mysql backend slows down the integration tests. My approach would be to use `derby` instead of mysql to speed things up - but I havent gotten around to take a look at it.

Anyway best regards and thanks in advance for the feedback.

**Comment by marvinlanhenke** [comment]
> In short, submit a PR and then we can polish it together.

Sounds good to me - looking forward to it.

Best regards.

**Comment by marvinlanhenke** [comment]
@Xuanwo @liurenjie1024 
I'm currently looking at the table operations in particular `create_table` and I'm having some trouble understanding the handling of the metadata. Looking at the `REST-Catalog` implementation, I don't see any metadata.json actually being written (physically)?

Is this correct or am I missing something? Can I also assume we don't have some `MetadataFileWriter` yet? 

So what would be the scope or the responsibility of the HMS Catalog Impl? 
In the Python impl for example, the metadata.json is written before the table is created in the catalog https://github.com/apache/iceberg-python/blob/main/pyiceberg/catalog/hive.py#L309

**Comment by liurenjie1024** [comment]
Hi, @marvinlanhenke I think currently we can leave `create_table`  unimplemented since it has some dependencies, such as transaction api and table metadata builder, otherwise it's currently not practical.

**Comment by marvinlanhenke** [comment]
> [...] since it has some dependencies, such as transaction api and table metadata builder, otherwise it's currently not practical.
Makes sense to me.

So I'd leave `create_table`, `load_table`, `update_table` and `drop_table` as unimplemented, since they all depend on those dependencies you mentioned - or they need create_table; so those other methods can be tested.

@liurenjie1024 
I think it'd be best to create another issue that sketches out the next steps (dependencies) for the upcomming week or two? (Andrew Lamb does this in Datafusion and it helps a lot coordinating the effort).

**Comment by liurenjie1024** [comment]
Hi, @marvinlanhenke Thanks for the suggestion, I think creating tracking issue and clarifying their dependency would be quite helpful, let me create one for 0.3 release. I'm not sure if an issue for upcoming week or two helps.

**Comment by liurenjie1024** [comment]
The left parts are blocked by #250

**Comment by marvinlanhenke** [comment]
> The left parts are blocked by #250

thank you so much for setting up the projects and preparing release 0.3.0 - I like the structure.

For this issue - I'll try to push part 2 this week where I setup as much as I can do without the `TableMetadataBuilder` - meaning all the table operations are implemented / including validations so the only part left, will be the creation of the `TableMetadata` and the actual Result<Table>.

**Comment by liurenjie1024** [comment]
@marvinlanhenke Thanks for the work. I'll also try to work on setting up `TableMetadataBuilder` this week.

**Comment by marvinlanhenke** [comment]
@liurenjie1024 
just a quick update: I've implemented most of the stuff I can do without the `TableMetadataBuilder`. Also you can assign this issue to me, if you'd like.

**Comment by liurenjie1024** [comment]
@marvinlanhenke Sure, thanks!

**Comment by marvinlanhenke** [comment]
@liurenjie1024 
I'm done with the implementation and ready to push the PR.
However, I think it would be great to review #272 first, since it blocked my implementation / serialization of metadata. Thus, I made the changes. Let me know what you think

**Comment by liurenjie1024** [comment]
@marvinlanhenke The  pr looks great, thanks!
